### DISCUSSION TITLE
PackFontRanges() creates a sigsegv crash

### DISCUSSION THREAD

**Discussion by heysokam** [nothings/stb#1377] [Q&A] [answered]
**Describe the bug**
On file: `stb_truetype.h`
Before execution of line #4118, access to `spc->pixels` works normally
After execution of the loop, the address stops being accesible, and code that needs to access the pixel buffer just segfaults the application.
```c
STBTT_DEF int stbtt_PackFontRanges(stbtt_pack_context *spc, const unsigned char *fontdata, int font_index, stbtt_pack_range *ranges, int num_ranges)
{
   stbtt_fontinfo info;
   int i,j,n, return_value = 1;
   //stbrp_context *context = (stbrp_context *) spc->pack_info;
   stbrp_rect    *rects;

// <---- spc->pixels memory is readable correctly, and zeroed as expected
   // flag all characters as NOT packed
   for (i=0; i < num_ranges; ++i)
      for (j=0; j < ranges[i].num_chars; ++j)
         ranges[i].chardata_for_range[j].x0 =
         ranges[i].chardata_for_range[j].y0 =
         ranges[i].chardata_for_range[j].x1 =
         ranges[i].chardata_for_range[j].y1 = 0;
// <---- spc->pixels memory is not accessible anymore
   n = 0;
```
The line that crashes with SIGSEGV is at #3192
```c
result->pixels[j*result->stride + i] = (unsigned char) m;
```
Code in between this line, and the previously mentioned for loop, works correctly without crashing.

**To Reproduce**
- Follow the Font Packing documentation in stb_truetype
- Try to pack font `LiberationMono-Bold.ttf`, with charset range `0x20 to 0x7F`
- Compile on linux, with `gcc -ggdb`
- Execution will end with a Segmentation Fault, during the call to `stbtt_PackFontRange`

**Expected behavior**
If the spc->pixels buffer is accessible correctly right before the loop to flag all characters as not packed, code that doesn't modify that pixel buffer shouldn't break access to the buffer either.

**Screenshots**
![image](https://user-images.githubusercontent.com/7308253/190914306-b966417a-cb8d-4940-a99e-795a7f167fa3.png)

**Comment by nothings** [comment]
The loop in question only processes `ranges`. `ranges` is a user-supplied array of structs. It sounds like you haven't initialized `ranges` correctly in your code.

**Comment by heysokam** [comment]
As mentioned on "steps to reproduce", I'm using `stbtt_PackFontRange`, not `stbtt_PackFontRanges` _(note the missing S)_ which generates such `ranges` struct from the function inputs `int first_unicode_codepoint_in_range` and `int num_chars_in_range` in `stb_truetype.h` from lines #4150 to #4155, right before calling `stbtt_PackFontRanges` with such self generated data

As said, the value for first is `0x20` and the value for num is `0x7F - 0x20`, which resolves to `95`. This covers all ascii printable characters, starting from space and ending in tilde. I checked with the debugger, and the integer values supplied to PackRange are indeed those at runtime.

Are these incorrect integer values to pass to the function?
![image](https://user-images.githubusercontent.com/7308253/190946746-a454a2a4-2544-4bfd-bb24-9fd0b982b90c.png)

**Comment by heysokam** [comment]
I gathered a bit more data.  
The loop breaks the `spc->pixels` address right before line #4120, when the value of `j` becomes `3`. So assuming this happens during the value assignment of the previous iteration, when `j=2`
Stepping through the loop, I noticed that the contents of `spc->nodes` are also being updated. I don't know if this is expected behavior, but noting it just in case.

**Comment by nothings** [comment]
I can only guess because you're not actually showing YOUR code which is where the bug lies.

You pass in `chardata_for_range` to `stbtt_PackFontRange`. Because you're asking for 95 characters, `chardata_for_range` needs to be an array of 95 elements. But probably you only have one, and the loop in question is overwriting random memory on the stack. Probably the spc->pxiels pointer is different after the loop than it was before the loop, because it got overwritten.

**Comment by heysokam** [comment, marked as answer]
Been trying to figure out how to use `stb_truetype.h` to create a font bitmap atlas.  
This is the code I have so far:
https://gist.github.com/heysokam/0270e3bcc36c6c0d537b476d5862dbc2#file-fontatlas-c-L291
As you can see, I haven't generated the `chardata_for_range` array, I just created an empty `stbtt_packedchar` struct, since I didn't know what to do with it.  

As far as my understanding reaches (first year into programming) this data structure contains information that is stored in the font file itself?
But I don't know how to generate such data.  

Is there existing documentation somewhere about this that I missed?  
If not, how can one create the required array?

**Reply by nothings** [reply]
It is an ambiguity of C that a function prototype with a pointer parameter like `stbtt_packedchar *chardata_for_range` may either expect a single struct or an array of structs, an ambiguity that you must resolve by reading the documentation.

In this case it expects an array of structs, which you could try to declare as:

```
PackChar fontPack[charNum];
```

However, in most versions of C this won't actually work because of how `const` works, so you need to either use a #define instead, or just directly declare the size you need (which isn't future-proof):

```
PackChar fontPack[95];
```

You don't need to initialize it beyond that, as it is the job of stbtt_PackFontRange to fill it out.

**Comment by heysokam** [comment]
That worked, thank you. I always fail at the smallest things possible :pensive:

Goal is to use this in a game engine, but such engine doesn't render quads, and also asking for render calls from the virtual machine overcomplicates things by a million (its not coded to allow that at all).
So aiming to just write a TGA file (the default format expected by the engine), that I can then cache directly using the pre-existing engine methods for that task.

How do you access the packed bitmap buffer, after the call to PackFontRanges has ended?
The documentation mentions `stbtt_GetPackedQuad`, but is this function usable to generate a bitmap image file directly from code?

**Reply by nothings** [reply]
The bitmap buffer is the same one you created, `atlasBit`. The info needed to indicate where each character is in the pixel buffer is in `fontPack`. `stbtt_GetPackedQuad ` is irrelevant, it's a convenience function for rendering quads which is irrelevant if you're not rendering quads.
